src/machine/usb/adc/midi/midi.go

```diff
@@ -53,24 +54,40 @@ func newMidi() *midi {
 				Index:     usb.MIDI_ENDPOINT_IN,
 				IsIn:      true,
 				Type:      usb.ENDPOINT_TYPE_BULK,
-				TxHandler: m.Handler,
+				TxHandler: m.TxHandler,
 			},
 		},
 		[]usb.SetupConfig{},
 	)
 	return m
 }
 
+// SetHandler is now deprecated, please use SetRxHandler().
 func (m *midi) SetHandler(rxHandler func([]byte)) {
+	m.SetRxHandler(rxHandler)
+}
+
+// SetRxHandler sets the handler function for incoming MIDI messages.
+func (m *midi) SetRxHandler(rxHandler func([]byte)) {
 	m.rxHandler = rxHandler
 }
 
+// SetTxHandler sets the handler function for outgoing MIDI messages.
+func (m *midi) SetTxHandler(txHandler func()) {
+	m.txHandler = txHandler
+}
+
 func (m *midi) Write(b []byte) (n int, err error) {
-	i := 0
-	for i = 0; i < len(b); i += 4 {
-		m.tx(b[i : i+4])
+	s, e := 0, 0
+	for s = 0; s < len(b); s += 4 {
+		e = s + 4
+		if e > len(b) {
+			e = len(b)
+		}
+
+		m.tx(b[s:e])
```

Comment:
I have no experience with MIDI, but just saw micro optimisation for this loop :)  
```suggestion
	e := 4
	for ; e < len(b); e += 4 {
		m.tx(b[e-4 : e])
	}
	if e < len(b)+4 {
		m.tx(b[e-4:])
	}
	return len(b), nil
```

Reply:
Thanks for the suggestion. @HattoriHanzo031 
Perhaps both are good codes.
I think the original code does not need to be changed as the loops look much clearer.

Reply:
@sago35 performance gain is probably negligible so if you think original code is more clear then it is best to leave it as is